basic-steps/src/main/java/org/jenkinsci/plugins/workflow/steps/TimeoutStepExecution.java

```diff
@@ -22,26 +29,29 @@ public boolean start() throws Exception {
                 .withCallback(new Callback())
                 .withDisplayName(null)  // hide the body block
                 .start();
-        setupTimer();
+        long now = step.getUnit().convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+        end = now + step.getTime();
```

Comment:
I think you got this backwards. The `convert` call should be on `step.time`, not `System.currentTimeMillis`.